File: Opc.Ua.RestfulApi/Node.cs
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace Opc.Ua.RestfulApi
{
    [JsonConverter(typeof(JsonStringEnumConverter))]
    public enum NodeClass
    {
        Object,
        Variable,
        Method,
        ObjectType,
        VariableType,
        ReferenceType,
        DataType,
        View
    }

    public class Node
    {
        [JsonPropertyName("nodeId")]
        public string NodeId { get; set; } = string.Empty;

        [JsonPropertyName("nodeClass")]
        public NodeClass NodeClass { get; set; }

        [JsonPropertyName("browseName")]
        public string BrowseName { get; set; } = string.Empty;

        [JsonPropertyName("displayName")]
        public LocalizedText DisplayName { get; set; } = new LocalizedText();

        [JsonPropertyName("description")]
        public LocalizedText? Description { get; set; }

        [JsonPropertyName("writeMask")]
        public int? WriteMask { get; set; }

        [JsonPropertyName("userWriteMask")]
        public int? UserWriteMask { get; set; }

        [JsonPropertyName("value")]
        public object? Value { get; set; }

        [JsonPropertyName("dataType")]
        public string? DataType { get; set; }

        [JsonPropertyName("dataTypeName")]
        public string? DataTypeName { get; set; }

        [JsonPropertyName("valueRank")]
        public int? ValueRank { get; set; }

        [JsonPropertyName("arrayDimensions")]
        public List<int>? ArrayDimensions { get; set; }

        [JsonPropertyName("accessLevel")]
        public int? AccessLevel { get; set; }

        [JsonPropertyName("userAccessLevel")]
        public int? UserAccessLevel { get; set; }

        [JsonPropertyName("minimumSamplingInterval")]
        public double? MinimumSamplingInterval { get; set; }

        [JsonPropertyName("historizing")]
        public bool? Historizing { get; set; }

        [JsonPropertyName("executable")]
        public bool? Executable { get; set; }

        [JsonPropertyName("userExecutable")]
        public bool? UserExecutable { get; set; }

        [JsonPropertyName("isAbstract")]
        public bool? IsAbstract { get; set; }

        [JsonPropertyName("designToolOnly")]
        public bool? DesignToolOnly { get; set; }

        [JsonPropertyName("isProperty")]
        public bool? IsProperty { get; set; }

        [JsonPropertyName("symmetric")]
        public bool? Symmetric { get; set; }

        [JsonPropertyName("inverseName")]
        public LocalizedText? InverseName { get; set; }

        [JsonPropertyName("containsNoLoops")]
        public bool? ContainsNoLoops { get; set; }

        [JsonPropertyName("eventNotifier")]
        public int? EventNotifier { get; set; }

        [JsonPropertyName("typeDefinition")]
        public string? TypeDefinition { get; set; }

        [JsonPropertyName("typeDefinitionName")]
        public string? TypeDefinitionName { get; set; }

        [JsonPropertyName("modellingRule")]
        public string? ModellingRule { get; set; }

        [JsonPropertyName("superTypeId")]
        public string? SuperTypeId { get; set; }

        [JsonPropertyName("superTypeIds")]
        public List<string>? SuperTypeIds { get; set; }

        [JsonPropertyName("parentNodeId")]
        public string? ParentNodeId { get; set; }

        /// <summary>
        /// The namespace URI the node lives in (its NodeId's "nsu=" part, or the core
        /// namespace for a bare NodeId). Clients render "[ModelName]:" prefixes from it.
        /// </summary>
        [JsonPropertyName("modelUri")]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string? ModelUri { get; set; }

        [JsonPropertyName("referenceType")]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string? ReferenceType { get; set; }

        [JsonPropertyName("referenceTypeId")]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string? ReferenceTypeId { get; set; }

        [JsonPropertyName("hasNoSubtypes")]
        public bool? HasNoSubtypes { get; set; }

        [JsonPropertyName("hasNoChildren")]
        public bool? HasNoChildren { get; set; }

        [JsonPropertyName("isInherited")]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public bool? IsInherited { get; set; }

        [JsonPropertyName("sourceTypeNodeId")]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string? SourceTypeNodeId { get; set; }

        [JsonPropertyName("isOverride")]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public bool? IsOverride { get; set; }

        [JsonPropertyName("dataTypeForm")]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string? DataTypeForm { get; set; }

        [JsonPropertyName("documentation")]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string? Documentation { get; set; }

        /// <summary>
        /// The conformance units this node belongs to — the <c>&lt;Category&gt;</c> elements of
        /// the NodeSet XML. Named for the XML element rather than the concept so the mapping
        /// stays obvious; omitted when the node names none.
        /// </summary>
        [JsonPropertyName("category")]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public List<string>? Category { get; set; }
    }
}

File: nodeseteditor.client/src/model/Node.ts
import type { LocalizedText } from './WorkspaceDescription';

/** Get the plain display name without namespace prefix */
export function getNodePlainName(node: { displayName?: LocalizedText; browseName?: string; nodeId?: string }): string {
   const name = node.displayName?.text ?? '';
   if (name) return name;
   const browseName = node.browseName ?? '';
   // Extract plain name from "nsu=URI;Name" format
   const semiIdx = browseName.indexOf(';');
   if (semiIdx >= 0) return browseName.substring(semiIdx + 1);
   // Fallback for "ModelName:Name" format
   const colonIdx = browseName.indexOf(':');
   if (colonIdx > 0) return browseName.substring(colonIdx + 1);
   return browseName || node.nodeId || '';
}

export interface Node {
   nodeId: string;
   nodeClass: string;
   browseName: string;
   displayName: LocalizedText;
   description?: LocalizedText;
   isAbstract?: boolean;
   // ReferenceType only. A symmetric reference reads the same in both
   // directions and therefore carries no inverseName.
   symmetric?: boolean;
   inverseName?: LocalizedText;
   // Top-level Object/Variable only: node exists only in the design tool
   // (no children/references, instantiation skipped). Set at creation only.
   designToolOnly?: boolean;
   // Variable whose TypeDefinition is PropertyType (i=68) or a subtype. A
   // Property may not be the source of hierarchical references (no children).
   isProperty?: boolean;
   superTypeId?: string;
   superTypeIds?: string[];
   parentNodeId?: string;
   hasNoSubtypes?: boolean;
   hasNoChildren?: boolean;
   /**
    * The subtype walk hit the requested depth here, so this node's subtypes are NOT in the
    * response even though it has some — re-browse from this node when it is expanded. Absent
    * means the response already carries whatever subtypes it has.
    */
   subtypesTruncated?: boolean;
   modelUri?: string;
   referenceType?: string;
   referenceTypeId?: string;
   typeDefinition?: string;
   typeDefinitionName?: string;
   modellingRule?: string;
   dataType?: string;
   dataTypeName?: string;
   valueRank?: number;
   arrayDimensions?: number[];
   isInherited?: boolean;
   sourceTypeNodeId?: string;
   isOverride?: boolean;
   dataTypeForm?: string;
   value?: unknown;
   documentation?: string;
   /**
    * Conformance units the node belongs to, one per entry — the <Category>
    * elements of the NodeSet XML.
    */
   category?: string[];
}

export interface ReferenceDescription {
   referenceTypeId: string;
   referenceTypeName?: string;
   isForward: boolean;
   targetNodeId: string;
   targetBrowseName?: string;
   targetDisplayName?: LocalizedText;
   targetNodeClass?: string;
   /**
    * True when this reference is the structural parent link for one of its
    * endpoints (Node table's ParentNodeId+ReferenceTypeId, or SuperTypeId
    * via HasSubtype for type nodes). The references view hides the delete
    * affordance for these — re-parenting/deleting the child is the only
    * way to remove them.
    */
   isCanonicalParent?: boolean;
}

export interface TreeNode extends Node {
   children: TreeNode[];
}

/** Build a tree from a flat list of nodes using superTypeId or parentNodeId. */
export function buildTree(nodes: Node[], rootId: string): TreeNode[] {
   const childrenMap = new Map<string, TreeNode[]>();

   for (const node of nodes) {
      const parentId = node.superTypeId ?? node.parentNodeId ?? rootId;
      const treeNode: TreeNode = { ...node, children: [] };

      if (!childrenMap.has(parentId)) {
         childrenMap.set(parentId, []);
      }
      childrenMap.get(parentId)!.push(treeNode);
   }

   function attachChildren(treeNode: TreeNode): void {
      const kids = childrenMap.get(treeNode.nodeId);
      if (kids) {
         treeNode.children = kids;
         for (const kid of kids) attachChildren(kid);
      }
   }

   const roots = childrenMap.get(rootId) ?? [];
   for (const root of roots) attachChildren(root);
   return roots;
}
